### PR TITLE
Optimizations for SubgroupShuffleXor() function

### PR DESCRIPTION
**Pull request by yjaelex**
LLPC will issue WaterfallLoop for this function in some cases.
Add optimizations for this function.
1. issue dpp_mov for some simple quad/row shuffle cases;
2. for maskValue <32, issue ds_permlane_x16 if supported or ds_swizzle;
3. then call SubgroupShuffle, which may issue waterfallloops to handle
complex cases.

### CODE REVIEW

File: lgc/builder/BuilderImpl.cpp [hunk 1]
```diff
@@ -81,6 +81,12 @@ bool BuilderImplBase::supportDpp() const {
   return getPipelineState()->getTargetInfo().getGfxIpVersion().major >= 8;
 }
 
+// =====================================================================================================================
```
Comment (trenouf) [inline lgc/builder/BuilderImpl.cpp:84]: We need to get rid of these "support*" functions and express them instead as target features in TargetInfo. But adding this one here for now is fine. :-)

File: lgc/builder/SubgroupBuilder.cpp [hunk 2]
```diff
@@ -317,9 +317,101 @@ Value *SubgroupBuilder::CreateSubgroupShuffle(Value *const value, Value *const i
...
+        }
+      }
+    }
+  }
+
+  if (maskValue < 32) {
+    if (supportDpp() && canOptimize)
+      return createDppMov(value, dppCtrl, 0xF, 0xF, true);
+    else if (supportPermLaneDpp() && (maskValue >= 16)) {
+      static const unsigned laneSelBits[16][2] = {
```
Comment (amdrexu) [inline lgc/builder/SubgroupBuilder.cpp:401]: laneSelBits -> LaneSelBits
Static variable uses uppercase-initial.

File: lgc/builder/SubgroupBuilder.cpp [hunk 3]
```diff
@@ -317,9 +317,101 @@ Value *SubgroupBuilder::CreateSubgroupShuffle(Value *const value, Value *const i
...
+          {0x76543210, 0xfedcba98}, {0x67452301, 0xefcdab89}, {0x54761032, 0xdcfe98ba}, {0x45670123, 0xcdef89ab},
```
Comment (amdrexu) [inline lgc/builder/SubgroupBuilder.cpp:402]: Use uppercase for hex number for this table.

File: lgc/builder/BuilderImpl.h [hunk 4]
```diff
@@ -721,18 +724,98 @@ class SubgroupBuilder : virtual public BuilderImplBase {
...
+    DppRowRr8 = 0x128,
+    DppRowRr9 = 0x129,
+    DppRowRr10 = 0x12A,
+    DppRowRr11 = 0x12B,
+    DppRowRr12 = 0x12C,
+    DppRowRr13 = 0x12D,
+    DppRowRr14 = 0x12E,
+    DppRowRr15 = 0x12F,
+
+    // WfSl and WfSr are not available on gfx10+.
```
Comment (amdrexu) [inline lgc/builder/BuilderImpl.h:775]: gfx10+ -> GFX10+

File: lgc/builder/BuilderImpl.h [hunk 5]
```diff
@@ -721,18 +724,98 @@ class SubgroupBuilder : virtual public BuilderImplBase {
...
     DppWfSl1 = 0x130,
     DppWfSr1 = 0x138,
+
     DppRowMirror = 0x140,
     DppRowHalfMirror = 0x141,
+
+    // RowBcast modes are not available on gfx10+.
```
Comment (amdrexu) [inline lgc/builder/BuilderImpl.h:782]: The same as above

File: lgc/builder/BuilderImpl.h [hunk 6]
```diff
@@ -721,18 +724,98 @@ class SubgroupBuilder : virtual public BuilderImplBase {
...
     DppRowBcast15 = 0x142,
     DppRowBcast31 = 0x143,
+
+    // RowXmask and RowShare modes are only available on gfx10+.
```
Comment (amdrexu) [inline lgc/builder/BuilderImpl.h:786]: The same as above

**Comment by amdvlk-admin** [comment]
### Test summary for commit 01f6b58fdb3ecbef6e5a0f00b5aa68e5bc9dbf97
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev e2f8eb0ffe21751b300995e795965d37e8d6dd67</li>
<li>SPVGEN: dev e843618d71c8263030551e7692b7683f20f627fe</li>
<li>XGL: dev 2c44bbc7b58efba12c501f71f5b268f83c1bdad5</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev f02738a5cf6c158c8c5b25a5048951f3059cabe2</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 958/958 (100.0%)</li>
<li>Failed: 0/958 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103142/103142 (100.0%)</li>
<li>Failed: 0/103142 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 88873/88874 (100.0%)</li>
<li><details><summary>Failed: 1/88874 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.arithmetic.framebuffer.subgroupor_i64vec3_vertex
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit bc42fbb1453155b8bedbcc0c52ab129db130cc2f
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev e2f8eb0ffe21751b300995e795965d37e8d6dd67</li>
<li>SPVGEN: dev e843618d71c8263030551e7692b7683f20f627fe</li>
<li>XGL: dev 2c44bbc7b58efba12c501f71f5b268f83c1bdad5</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev f02738a5cf6c158c8c5b25a5048951f3059cabe2</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 958/958 (100.0%)</li>
<li>Failed: 0/958 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103142/103142 (100.0%)</li>
<li>Failed: 0/103142 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 89053/89053 (100.0%)</li>
<li>Failed: 0/89053 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/BuilderImpl.cpp (hunk 1 above)
Comment (nhaehnle) [reply lgc/builder/BuilderImpl.cpp:84]: FWIW, I'm of different minds on this one.

On the one hand, it makes things more orthogonal and separated out. On the other hand, it causes us to forget what each hardware supports.

Comment (amdrexu) [reply lgc/builder/BuilderImpl.cpp:84]: A better way is to have all features encapsulated in target info class and target info provides a bunch of query functions (based on GFX IP) to get whatever we want to know.

**Pull request merged**